File: Claude-Project-06/Lacy_HMRN/scripts/annotate_de_probes.R
# =============================================================================
# Annotate Differentially Expressed Probes with Gene Symbols
# Uses GPL14951 platform annotation
# =============================================================================

cat("=============================================================\n")
cat("Annotating DE Probes with Gene Symbols\n")
cat("=============================================================\n\n")

library(data.table)

lacy_dir <- "C:/Users/ericp/OneDrive/Desktop/Claude-Projects/Claude-Project-06/Lacy_HMRN"
results_dir <- file.path(lacy_dir, "results")

# =============================================================================
# 1. Load DE Results
# =============================================================================

cat("Loading DE results...\n")
de_results <- read.csv(file.path(results_dir, "gcb_expression_by_stage.csv"),
                       stringsAsFactors = FALSE)
cat("Total probes:", nrow(de_results), "\n\n")

# =============================================================================
# 2. Download GPL14951 Annotation
# =============================================================================

cat("Downloading GPL14951 annotation...\n")

# Try to download the annotation file
gpl_url <- "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL14951&targ=self&view=data&form=text"

# Download to temp file
temp_file <- tempfile(fileext = ".txt")
tryCatch({
  download.file(gpl_url, temp_file, mode = "wb", quiet = TRUE)
  cat("Downloaded GPL annotation\n")
}, error = function(e) {
  cat("Failed to download:", e$message, "\n")
})

# Read and parse the annotation
if (file.exists(temp_file) && file.size(temp_file) > 0) {

  # Read lines to find the data table
  lines <- readLines(temp_file, warn = FALSE)

  # Find table start
  table_start <- grep("^!platform_table_begin", lines)
  table_end <- grep("^!platform_table_end", lines)

  if (length(table_start) > 0 && length(table_end) > 0) {
    # Extract table content
    table_lines <- lines[(table_start + 1):(table_end - 1)]

    # Parse as data frame
    gpl_annot <- fread(text = paste(table_lines, collapse = "\n"),
                       header = TRUE, sep = "\t", fill = TRUE)

    cat("Annotation columns:\n")
    print(names(gpl_annot))
    cat("\n")

    # Find relevant columns
    id_col <- grep("^ID$|^ID_REF$|^Probe_Id", names(gpl_annot), value = TRUE)[1]
    symbol_col <- grep("Symbol|GENE_SYMBOL|Gene_Symbol", names(gpl_annot),
                       ignore.case = TRUE, value = TRUE)[1]

    cat("ID column:", id_col, "\n")
    cat("Symbol column:", symbol_col, "\n\n")

    if (!is.na(id_col) && !is.na(symbol_col)) {
      # Create annotation mapping
      annot_map <- data.frame(
        Probe = gpl_annot[[id_col]],
        Gene_Symbol = gpl_annot[[symbol_col]],
        stringsAsFactors = FALSE
      )

      # Clean up gene symbols
      annot_map$Gene_Symbol[annot_map$Gene_Symbol == ""] <- NA

      cat("Annotation entries:", nrow(annot_map), "\n")
      cat("With gene symbols:", sum(!is.na(annot_map$Gene_Symbol)), "\n\n")
    }
  }
}

# If download failed, try to extract from series matrix
if (!exists("annot_map")) {
  cat("Trying to extract annotation from series matrix...\n")

  series_file <- file.path(lacy_dir, "GSE181063_series_matrix.txt.gz")
  con <- gzfile(series_file, "r")
  lines <- readLines(con, n = 500)  # Just first 500 lines for metadata
  close(con)

  # Look for gene symbol info
  gene_line <- grep("gene_symbol", lines, ignore.case = TRUE, value = TRUE)
  if (length(gene_line) > 0) {
    cat("Found gene symbol line in series matrix\n")
    print(substr(gene_line[1], 1, 200))
  }
}

# =============================================================================
# 3. Alternative: Use BioMart for annotation
# =============================================================================

if (!exists("annot_map") || sum(!is.na(annot_map$Gene_Symbol)) < 1000) {
  cat("\nUsing Illumina array annotation from series matrix...\n")

  # The series matrix has gene assignments - extract from full file
  series_file <- file.path(lacy_dir, "GSE181063_series_matrix.txt.gz")
  con <- gzfile(series_file, "r")
  all_lines <- readLines(con)
  close(con)

  # Look for platform annotation section
  platform_ref <- grep("!Sample_platform_id", all_lines, value = TRUE)
  cat("Platform reference:", substr(platform_ref[1], 1, 100), "\n")

  # Try to get gene info from sample characteristics
  gene_rows <- grep("!Sample_characteristics", all_lines, value = TRUE)
  cat("Found", length(gene_rows), "sample characteristic lines\n")
}

# =============================================================================
# 4. Manual annotation lookup for top hits
# =============================================================================

cat("\n=============================================================\n")
cat("Manual Annotation for Top DE Probes\n")
cat("=============================================================\n\n")

# Top probes to annotate (FDR < 0.1)
top_probes <- de_results[de_results$FDR < 0.1, ]
cat("Probes with FDR < 0.1:", nrow(top_probes), "\n\n")

# If we have annotation, merge
if (exists("annot_map") && nrow(annot_map) > 0) {

  de_annotated <- merge(de_results, annot_map, by = "Probe", all.x = TRUE)
  de_annotated <- de_annotated[order(de_annotated$P_value), ]

  cat("Annotated probes:", sum(!is.na(de_annotated$Gene_Symbol)), "\n\n")

  # Show top hits
  cat("Top 30 DE Genes by Stage (GCB-DLBCL):\n")
  cat("=====================================\n\n")

  top30 <- head(de_annotated[!is.na(de_annotated$Gene_Symbol), ], 30)
  print(top30[, c("Gene_Symbol", "Log2FC", "P_value", "FDR", "Direction")])

  # Separate by direction
  cat("\n\nTop Limited-high Genes (higher in Stage I-II):\n")
  limited_high <- de_annotated[de_annotated$Direction == "Limited-high" &
                                !is.na(de_annotated$Gene_Symbol), ]
  print(head(limited_high[, c("Gene_Symbol", "Log2FC", "P_value", "FDR")], 20))

  cat("\nTop Advanced-high Genes (higher in Stage III-IV):\n")
  advanced_high <- de_annotated[de_annotated$Direction == "Advanced-high" &
                                 !is.na(de_annotated$Gene_Symbol), ]
  print(head(advanced_high[, c("Gene_Symbol", "Log2FC", "P_value", "FDR")], 20))

  # Save annotated results
  write.csv(de_annotated, file.path(results_dir, "gcb_de_genes_by_stage_annotated.csv"),
            row.names = FALSE)
  cat("\nSaved: gcb_de_genes_by_stage_annotated.csv\n")

  # Summary statistics
  cat("\n=============================================================\n")
  cat("SUMMARY\n")
  cat("=============================================================\n\n")

  sig_genes <- de_annotated[de_annotated$FDR < 0.1 & !is.na(de_annotated$Gene_Symbol), ]
  cat("Significant genes (FDR < 0.1):", nrow(sig_genes), "\n")
  cat("  Limited-high:", sum(sig_genes$Direction == "Limited-high"), "\n")
  cat("  Advanced-high:", sum(sig_genes$Direction == "Advanced-high"), "\n")

} else {
  cat("Could not load annotation - saving probe-only results\n")

  # Still save with direction summary
  cat("\nTop 30 DE Probes (unannotated):\n")
  print(head(de_results[, c("Probe", "Log2FC", "P_value", "FDR", "Direction")], 30))
}

cat("\n=============================================================\n")
cat("ANNOTATION COMPLETE\n")
cat("=============================================================\n")

File: Claude-Project-06-v2/Lacy_HMRN/scripts/annotate_de_probes.py
"""
Annotate Differentially Expressed Probes with Gene Symbols
Uses GPL14951 (Illumina DASL) platform annotation
"""

import pandas as pd
import urllib.request
import gzip
import io
import os

print("=" * 60)
print("Annotating DE Probes with Gene Symbols")
print("=" * 60 + "\n")

lacy_dir = "C:/Users/ericp/OneDrive/Desktop/Claude-Projects/Claude-Project-06/Lacy_HMRN"
results_dir = os.path.join(lacy_dir, "results")

# =============================================================================
# 1. Load DE Results
# =============================================================================

print("Loading DE results...")
de_results = pd.read_csv(os.path.join(results_dir, "gcb_expression_by_stage.csv"))
print(f"Total probes: {len(de_results)}")
print(f"FDR < 0.1: {(de_results['FDR'] < 0.1).sum()}")
print(f"P < 0.01: {(de_results['P_value'] < 0.01).sum()}\n")

# =============================================================================
# 2. Download GPL14951 Annotation
# =============================================================================

print("Downloading GPL14951 annotation...")

# Use the SOFT format which is more parseable
gpl_url = "https://ftp.ncbi.nlm.nih.gov/geo/platforms/GPL14nnn/GPL14951/soft/GPL14951_family.soft.gz"

try:
    with urllib.request.urlopen(gpl_url, timeout=120) as response:
        print("Downloaded GPL annotation file")

        # Decompress and parse
        content = gzip.decompress(response.read()).decode('utf-8', errors='ignore')
        lines = content.split('\n')

        # Find the table section
        table_start = None
        table_end = None
        for i, line in enumerate(lines):
            if line.startswith("!platform_table_begin"):
                table_start = i + 1
            elif line.startswith("!platform_table_end"):
                table_end = i
                break

        if table_start and table_end:
            print(f"Found annotation table: lines {table_start} to {table_end}")

            # Parse table
            table_content = '\n'.join(lines[table_start:table_end])
            annot_df = pd.read_csv(io.StringIO(table_content), sep='\t',
                                   low_memory=False, on_bad_lines='skip')

            print(f"Annotation columns: {list(annot_df.columns)}")
            print(f"Annotation entries: {len(annot_df)}")

            # Find ID and Symbol columns
            id_cols = [c for c in annot_df.columns if c.upper() in ['ID', 'PROBE_ID', 'ID_REF']]
            symbol_cols = [c for c in annot_df.columns if 'SYMBOL' in c.upper() or 'GENE' in c.upper()]

            print(f"\nID columns found: {id_cols}")
            print(f"Symbol columns found: {symbol_cols}")

except Exception as e:
    print(f"Download failed: {e}")
    annot_df = None

# =============================================================================
# 3. Alternative: Check if annotation already exists locally
# =============================================================================

if annot_df is None or len(annot_df) == 0:
    print("\nChecking for local annotation...")

    # Try to see if there's already an annotation file
    local_annot = os.path.join(lacy_dir, "GPL14951_annotation.csv")
    if os.path.exists(local_annot):
        annot_df = pd.read_csv(local_annot)
        print(f"Loaded local annotation: {len(annot_df)} entries")

# =============================================================================
# 4. Merge Annotation with DE Results
# =============================================================================

print("\n" + "=" * 60)
print("Merging Annotation with DE Results")
print("=" * 60 + "\n")

if annot_df is not None and len(annot_df) > 0:
    # Find the right columns
    id_col = None
    symbol_col = None

    for col in annot_df.columns:
        if col.upper() == 'ID':
            id_col = col
        if 'SYMBOL' in col.upper():
            symbol_col = col

    if id_col and symbol_col:
        # Create mapping
        annot_map = annot_df[[id_col, symbol_col]].copy()
        annot_map.columns = ['Probe', 'Gene_Symbol']
        annot_map = annot_map.dropna(subset=['Gene_Symbol'])
        annot_map = annot_map[annot_map['Gene_Symbol'] != '']

        print(f"Probes with gene symbols: {len(annot_map)}")

        # Merge with DE results
        de_annotated = de_results.merge(annot_map, on='Probe', how='left')

        # Sort by p-value
        de_annotated = de_annotated.sort_values('P_value')

        annotated_count = de_annotated['Gene_Symbol'].notna().sum()
        print(f"DE probes annotated: {annotated_count} / {len(de_annotated)}")

        # =============================================================
        # Show Results
        # =============================================================

        print("\n" + "=" * 60)
        print("TOP GENES BY STAGE ASSOCIATION (GCB-DLBCL)")
        print("=" * 60 + "\n")

        # Filter to annotated probes
        annotated = de_annotated[de_annotated['Gene_Symbol'].notna()].copy()

        # Top 30 overall
        print("Top 30 Differentially Expressed Genes:")
        print("-" * 50)
        top30 = annotated.head(30)
        for i, row in top30.iterrows():
            direction = "↑" if row['Direction'] == "Advanced-high" else "↓"
            fdr_str = f"FDR={row['FDR']:.4f}" if row['FDR'] < 0.1 else f"p={row['P_value']:.2e}"
            print(f"  {row['Gene_Symbol']:15s} {direction} Adv  {fdr_str:20s} FC={row['Log2FC']:.3f}")

        # Significant genes by direction
        sig = annotated[annotated['FDR'] < 0.1]

        print(f"\n\nSignificant Genes (FDR < 0.1): {len(sig)}")
        print("-" * 50)

        limited_high = sig[sig['Direction'] == 'Limited-high'].sort_values('P_value')
        advanced_high = sig[sig['Direction'] == 'Advanced-high'].sort_values('P_value')

        print(f"\nHIGHER in Limited Stage (I-II): {len(limited_high)} genes")
        if len(limited_high) > 0:
            print("-" * 40)
            for i, row in limited_high.head(20).iterrows():
                print(f"  {row['Gene_Symbol']:15s} FDR={row['FDR']:.4f}  FC={row['Log2FC']:.3f}")

        print(f"\nHIGHER in Advanced Stage (III-IV): {len(advanced_high)} genes")
        if len(advanced_high) > 0:
            print("-" * 40)
            for i, row in advanced_high.head(20).iterrows():
                print(f"  {row['Gene_Symbol']:15s} FDR={row['FDR']:.4f}  FC={row['Log2FC']:.3f}")

        # Look for pathway-related genes
        print("\n" + "=" * 60)
        print("PATHWAY ANALYSIS")
        print("=" * 60 + "\n")

        # Egress/Retention pathway
        egress_genes = ['S1PR1', 'S1PR2', 'GNA13', 'RHOA', 'P2RY8', 'CXCR4', 'SGK1',
                       'GNAI2', 'RAC2', 'ARHGEF1', 'ARHGAP25', 'S1PR1', 'FOXO1']

        pathway_hits = annotated[annotated['Gene_Symbol'].isin(egress_genes)]
        print(f"Egress/Retention Pathway Genes in DE results: {len(pathway_hits)}")
        if len(pathway_hits) > 0:
            for i, row in pathway_hits.iterrows():
                sig_str = "*" if row['FDR'] < 0.1 else ""
                print(f"  {row['Gene_Symbol']:15s} p={row['P_value']:.4f} FC={row['Log2FC']:.3f} {sig_str}")

        # B-cell/lymphoma related
        bcell_genes = ['BCL2', 'BCL6', 'MYC', 'CD19', 'CD20', 'MS4A1', 'PAX5',
                      'IRF4', 'PRDM1', 'XBP1', 'CD38', 'CD27', 'AICDA']

        bcell_hits = annotated[annotated['Gene_Symbol'].isin(bcell_genes)]
        print(f"\nB-cell/Lymphoma Markers in DE results: {len(bcell_hits)}")
        if len(bcell_hits) > 0:
            for i, row in bcell_hits.iterrows():
                sig_str = "*" if row['FDR'] < 0.1 else ""
                print(f"  {row['Gene_Symbol']:15s} p={row['P_value']:.4f} FC={row['Log2FC']:.3f} {sig_str}")

        # Cell adhesion/migration
        migration_genes = ['ITGB1', 'ITGB2', 'ITGA4', 'ICAM1', 'VCAM1', 'CCR7',
                          'CXCR5', 'CCL19', 'CCL21', 'SELL', 'CD62L', 'VLA4']

        migration_hits = annotated[annotated['Gene_Symbol'].isin(migration_genes)]
        print(f"\nMigration/Adhesion Genes in DE results: {len(migration_hits)}")
        if len(migration_hits) > 0:
            for i, row in migration_hits.iterrows():
                sig_str = "*" if row['FDR'] < 0.1 else ""
                print(f"  {row['Gene_Symbol']:15s} p={row['P_value']:.4f} FC={row['Log2FC']:.3f} {sig_str}")

        # Save annotated results
        output_file = os.path.join(results_dir, "gcb_de_genes_by_stage_annotated.csv")
        de_annotated.to_csv(output_file, index=False)
        print(f"\n\nSaved: {output_file}")

        # Create summary table of significant genes
        if len(sig) > 0:
            sig_summary = sig[['Gene_Symbol', 'Log2FC', 'P_value', 'FDR', 'Direction']].copy()
            sig_summary.to_csv(os.path.join(results_dir, "gcb_de_significant_genes.csv"),
                              index=False)
            print(f"Saved: gcb_de_significant_genes.csv ({len(sig)} genes)")
    else:
        print(f"Could not find ID or Symbol columns. Available: {list(annot_df.columns)}")
else:
    print("No annotation available - showing probe IDs only")

    print("\nTop 30 DE Probes (FDR < 0.1 or top by p-value):")
    top = de_results.head(30)
    for i, row in top.iterrows():
        direction = "↑" if row['Direction'] == "Advanced-high" else "↓"
        print(f"  {row['Probe']:20s} {direction} Adv  FDR={row['FDR']:.4f}  FC={row['Log2FC']:.3f}")

print("\n" + "=" * 60)
print("ANNOTATION COMPLETE")
print("=" * 60)
